operator/pkg/awsprovider/iam/reconciler.go

```diff
@@ -0,0 +1,210 @@
+	})
+	if err != nil && !errors.IsIAMObjectDoNotExist(err) {
+		return fmt.Errorf("deleting instance profile, %w", err)
+	}
+
+	for _, policy := range desiredPolicies {
+		if _, err = c.iam.DetachRolePolicyWithContext(ctx, &iam.DetachRolePolicyInput{
+			PolicyArn: aws.String(policy),
+			RoleName:  aws.String(KitNodeRoleNameFor(controlPlane.ClusterName())),
+		}); err != nil {
+			return fmt.Errorf("detaching policy from role, %w", err)
+		}
+	}
+	_, err = c.iam.DeleteRoleWithContext(ctx, &iam.DeleteRoleInput{
+		RoleName: aws.String(KitNodeRoleNameFor(controlPlane.ClusterName())),
+	})
+	if err != nil && !errors.IsIAMObjectDoNotExist(err) {
+		return fmt.Errorf("deleting role, %w", err)
+	}
+	zap.S().Infof("[%s] Deleted IAM Role %v and instance profile",
+		controlPlane.ClusterName(), KitNodeRoleNameFor(controlPlane.ClusterName()))
+	return nil
+}
+
+func (c *Controller) getRole(ctx context.Context, controlPlane *apis.ControlPlane) (*role, error) {
+	roleOutput, err := c.iam.GetRoleWithContext(ctx, &iam.GetRoleInput{
+		RoleName: aws.String(KitNodeRoleNameFor(controlPlane.ClusterName())),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("getting iam role %v, %w", KitNodeRoleNameFor(controlPlane.ClusterName()), err)
+	}
+	return &role{iam: c.iam, Role: roleOutput.Role}, nil
+}
+
+func (c *Controller) createRole(ctx context.Context, controlPlane *apis.ControlPlane) (*role, error) {
+	roleOutput, err := c.iam.CreateRole(&iam.CreateRoleInput{
+		AssumeRolePolicyDocument: aws.String(assumeRolePolicyDocument),
+		Description:              aws.String("Role assumed by dataplane nodes created by KIT operated"),
+		RoleName:                 aws.String(KitNodeRoleNameFor(controlPlane.ClusterName())),
+		Tags:                     generateRoleTags(controlPlane.ClusterName()),
+	})
+	return &role{iam: c.iam, Role: roleOutput.Role}, err
+}
+
+type role struct {
+	iam *awsprovider.IAM
+	*iam.Role
+}
+
+func (r *role) addRoleToInstanceProfile(ctx context.Context, controlPlane *apis.ControlPlane) error {
+	profile, err := createInstanceProfile(ctx, r.iam, KitNodeInstanceProfileNameFor(controlPlane.ClusterName()))
+	if err != nil {
+		return fmt.Errorf("creating instance profile, %w", err)
+	}
+	if profile == nil {
+		return fmt.Errorf("instance profile is NIL")
+	}
+	for _, role := range profile.Roles {
+		if aws.StringValue(role.RoleName) == KitNodeRoleNameFor(controlPlane.ClusterName()) {
+			return nil
+		}
+	}
+	_, err = r.iam.AddRoleToInstanceProfile(&iam.AddRoleToInstanceProfileInput{
+		InstanceProfileName: aws.String(KitNodeInstanceProfileNameFor(controlPlane.ClusterName())),
+		RoleName:            aws.String(KitNodeRoleNameFor(controlPlane.ClusterName())),
+	})
+	return err
+}
+
+func (r *role) attachPolicy(ctx context.Context, policyARN, name string) error {
+	_, err := r.iam.AttachRolePolicy(&iam.AttachRolePolicyInput{
+		PolicyArn: aws.String(policyARN),
+		RoleName:  aws.String(KitNodeRoleNameFor(name)),
```

Comment:
function name sounds generic but this function is only good for `kitNodeROle` based on this line.

Reply:
update params to provide roleName instead of clusterName